### PR TITLE
Feature/#26 null data handling

### PR DESCRIPTION
**Pull request by ranajitbanerjee**
- Handle invalid types
- Add configuration to customize invalid value representation

### CODE REVIEW

File: packages/visual-group/src/encoder/cartesian-encoder.js [hunk 1]
```diff
@@ -75,7 +75,10 @@ export default class CartesianEncoder extends VisualEncoder {
             } else {
                 axesCreators.position = this.axisFrom()[type];
             }
-            geomCellAxes[axis] = generateAxisFromMap(axis, axisFields[i], axesCreators, axis === X ? xAxes : yAxes);
+            geomCellAxes[axis] = generateAxisFromMap(axis, axisFields[i], axesCreators, {
```
Comment (adotg) [inline packages/visual-group/src/encoder/cartesian-encoder.js]: Why is the axis created like this?
Use sanitizer to create a common data structure which then drives the creation of axis from `generateAxisFromMap`

File: packages/muze-axis/src/cartesian-axis/band-axis.js [hunk 2]
```diff
@@ -158,6 +167,6 @@ export default class BandAxis extends SimpleAxis {
      */
     invert (...value) {
         const values = value.map(d => this.scale().invert(d)) || [];
-        return value.length === 1 ? values[0] && values[0].toString() : values.map(d => d.toString());
+        return value.length === 1 ? values[0] : values.map(d => d);
```
[the remark this answers is not in the document]
Comment (ranajitbanerjee) [reply packages/muze-axis/src/cartesian-axis/band-axis.js]: `axis.invert(20)` returns single value
`axis.invert(20, 30, 40)` returns an array

File: packages/muze-axis/src/color-axis/color-axis.js [hunk 3]
```diff
@@ -128,12 +128,13 @@ export default class ColorAxis {
             const scale = this.scale();
             const range = scale.range ? scale.range() : null;
             const color = this._colorStrategy.value(range)(domainVal, scale, this.domain(), this.uniqueValues());
-
-            if (typeof color === 'string') {
-                const col = color.substring(color.indexOf('(') + 1, color.lastIndexOf(')')).split(/,\s*/);
-                return rgbToHsv(...col);
+            if (color) {
+                if (typeof color === 'string') {
```
Comment (UD-UD) [inline packages/muze-axis/src/color-axis/color-axis.js]: condition can merged in single if

File: packages/muze-axis/src/cartesian-axis/band-axis.js [hunk 4]
```diff
@@ -123,6 +123,15 @@ export default class BandAxis extends SimpleAxis {
         return this.axis().scale().domain();
     }
 
+    getTickFormatter (value) {
+        const { tickFormat } = value;
+
+        if (tickFormat) {
```
[the remark this answers is not in the document]
Comment (ranajitbanerjee) [reply packages/muze-axis/src/cartesian-axis/band-axis.js]: Actually it is doing a sanitization here. So changed the name to `sanitizeTickFormatter` method

File: packages/muze-axis/src/color-axis/color-axis.js (hunk 3 above)
Comment (ranajitbanerjee) [reply packages/muze-axis/src/color-axis/color-axis.js]: No, in this case it's not possible as there are 3 conditions color can be undefined or be a string or array

File: packages/muze-axis/src/cartesian-axis/band-axis.js [hunk 5]
```diff
@@ -88,10 +87,11 @@ export default class BandAxis extends SimpleAxis {
 
         if (domain && domain.length) {
```
[the remark this answers is not in the document]
Comment (adotg) [reply packages/muze-axis/src/cartesian-axis/band-axis.js]: And why is that?

File: packages/muze-axis/src/cartesian-axis/band-axis.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (adotg) [reply packages/muze-axis/src/cartesian-axis/band-axis.js]: Make it a consistent format.

File: packages/muze-axis/src/color-axis/color-axis.js (hunk 3 above)
Comment (UD-UD) [reply packages/muze-axis/src/color-axis/color-axis.js]: in that case it is advisable that the method call  
`this._colorStrategy.value(range)(domainVal, scale, this.domain(), this.uniqueValues());`

returns a result of single type.

File: packages/muze-utils/src/common-utils.js [hunk 6]
```diff
@@ -191,12 +198,14 @@ const getDomainFromData = (data, fields, fieldType) => {
...
-    let domain;
+    let domain = [];
     domains = domains.filter(dom => dom && dom.length);
-    if (fieldType === DimensionSubtype.CATEGORICAL) {
-        domain = domain = [].concat(...domains);
-    } else {
-        domain = [Math.min(...domains.map(d => d[0])), Math.max(...domains.map(d => d[1]))];
+    if (domains.length) {
+        if (fieldType === DimensionSubtype.CATEGORICAL) {
+            domain = domain = [].concat(...domains);
```
Comment (UD-UD) [inline packages/muze-utils/src/common-utils.js]: why assign multiple time ` domain = domain = [].concat(...domains);`

Comment (ranajitbanerjee) [reply packages/muze-utils/src/common-utils.js]: Changed

File: packages/muze-axis/src/cartesian-axis/band-axis.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ranajitbanerjee) [reply packages/muze-axis/src/cartesian-axis/band-axis.js]: This is done.Now invert method takes and returns a single value.

File: packages/muze-axis/src/color-axis/color-axis.js (hunk 3 above)
Comment (ranajitbanerjee) [reply packages/muze-axis/src/color-axis/color-axis.js]: Ok, this issue will be handled in #118

File: packages/muze-axis/src/cartesian-axis/band-axis.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (ranajitbanerjee) [reply packages/muze-axis/src/cartesian-axis/band-axis.js]: Ok, this will be taken care of in #118

**Pull request merged**